#+title: Project Outline
#+description: Exam Project "Mapping the reads from NGS data to the genome"
#+authors: Abdullah Cetinkaya, Niklas Horner, Philip Wolper

Link to GitHub repository: https://github.com/pwolper/read-mapping

***** Tasks to-do [0/4]
- [-] Added two pie charts to R code. Maybe still one missing showing percentage of reads for each number of hits. Improve coloring
- [ ] Fix the file path pointer accross the code.
- [ ] Clean up Rmd output of the results. Arrange charts with each other (ggarrange? Can we run the render process from python script?
- [ ] Write description of the project in the README. Maybe present using github website.

* Description
_“Mapping” the reads from NGS data to the genome_

- You have sequenced a genome with Illumina sequencing machine (single-end mode) and got a list of reads in FASTA file. For each read report the coordinate of the nucleotide where is has been mapped and to which strand it has been mapped.
- Please also keep in mind that DNA is double-stranded. In Fasta file of the genome you have only one strand reported, though the read can be mapped onto any of the strand. So please check both strands.
- Please provide the infographics showing how many reads have been mapped to the genome, how many have been mapped several times (how many times exactly?), how many haven’t been mapped.

* Heading + Project Infos
_Project directory structure_
#+begin_src shell
$ ls -l

total 16
drwxrwxr-x 2 philip philip 4096 Feb  4 14:12 code
drwxrwxr-x 2 philip philip 4096 Jan 18 09:03 data
drwxrwxr-x 3 philip philip 4096 Feb  4 14:11 output
-rw-rw-r-- 1 philip philip 2882 Feb  4 14:59 README.org
#+end_src

* Data

#+begin_src shell
-rw-rw-r-- 1 philip philip 4699745 Jan 27 15:37 Ecoli_genome.fasta
-rw-rw-r-- 1 philip philip   35890 Jan 18 09:02 reads_new.fasta
#+end_src

** Genome sequencing data
The genome data is required as /.fasta/ file format, and is found in the \/data/ folder. Parsing of this happens in mapping.py.

#+begin_src shell :results code
head data/Ecoli_genome.fasta
#+end_src

#+begin_src shell
>NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome
AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATATAGGCATAGCGCACAGAC
AGATAAAAATTACAGAGTACACAACATCCATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCACAGGTAACGGTGCGGGCTGACGCGTACAGGAAACACAGAAAAAAGCCCGCACCTGACAGTGCGGGCTTTTTTTTTCGACCAAAGG
TAACGAGGTAACAACCATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAACGTTTTCTGCGTGTTGCCGATATTCTGGAAAGCAATGCCAGGCAGGGGCAGGTGGCCACCGTCCTCTCTGCCCCCGCCAAAATCACCAACCACCTGGTG
GCGATGATTGAAAAAACCATTAGCGGCCAGGATGCTTTACCCAATATCAGCGATGCCGAACGTATTTTTGCCGAACTTTTGACGGGACTCGCCGCCGCCCAGCCGGGGTTCCCGCTGGCGCAATTGAAAACTTTCGTCGATCAGGAATTTGCCCAAATAA
AACATGTCCTGCATGGCATTAGTTTGTTGGGGCAGTGCCCGGATAGCATCAACGCTGCGCTGATTTGCCGTGGCGAGAAA
#+end_src


** Short sequencing reads
insert bash code

* How does the code work?
Write a decription of the code here...

** mapping.py
...
Source code blocks like this:

#+begin_src python
example = list()

for x in range(1,10):
    y = 3*x
    return("Return hello world!")

#+end_src

* Mapping output and Statistics
** python returns a .csv file
insert Excel screenshot

** Visulisation with R and ggplot2

insert tiff images to README.

* Resources

** NGS mapping from Tutorial for learning Genomics.
[[https://genomics.sschmeier.com/ngs-mapping/]]

** PieChart
https://r-graph-gallery.com/pie-plot.html

** Groups
Book feedback session:
[[https://docs.google.com/document/d/1gqQZ-vmIGHv3nXg-H3GatdHKmJ-UM38COClTAJtsNO0/edit]]

Chat to find groups:
[[https://docs.google.com/document/d/1RknEWydbXxivWt1V83t9NufH-o0nXl4EqAnhhL4Eb0Y/edit]]

Project and student list:
[[https://docs.google.com/document/d/1nLNA5moYD3z_DvdHuh5t_v5nju3Gx5ferpy24_2UbcI/edit]]
