# Visualisation of mapped sequencing reads

```{R message=F}
library(ggplot2)
library(dplyr)
library(purrr)

if (!require(ape)) {
  install.packages("ape")
  library(ape)
}


```

```{R}
genome <- read.dna("data/Ecoli_genome.fasta",format="fasta")
sequence <- as.character(genome)
genome_length <- length(sequence)
```

## Importing data from /output/*.csv
Import and formatting fo data to be used for vizualisation!
```{R}
mapped <- read.csv("output/2023-01-27_NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome.csv",sep = ";")

# mapped = mapped[50:60,]

mapped = mapped %>% mutate_at(vars(Sense, Antisense), ~map(strsplit(. ,split=","), ~map_int(.x, as.integer)))
mapped %>% str()

sense = unlist(mapped$Sense)
anti = unlist(mapped$Antisense)

sense %>% length()
anti %>% length()

reads <- data.frame(pos = c(sense, anti), strand = as.factor(rep(c("s","a"),times = c(length(sense),length(anti)))))

#pie charts
pie1 <- data.frame(Group = c("Sense", "Anti-Sense"), amount = c(length(sense),length(anti)))
pie1 <- pie1 %>%
  arrange(desc(Group)) %>%
  mutate(ypos = c(410, 905.5))

i = 1
pie2 <- data.frame("Group" =  c("No Hit", "Hit"), "value" = c(0,0))
while (i < length(mapped[,1])){
  i <- i+1
  if ((mapped[i,4]== 0) & (mapped[i,5] == 0)){
    pie2$value[1] <- pie2$value[1]+1
  }
  else{
    pie2$value[2] <- pie2$value[2]+1
  }
}
pie2$ypos <- c(pie2$value[1]/2, pie2$value[2]/2+pie2$value[1])
```

## Plotting mapped reads
```{R}

ggplot(reads, aes(x = pos, y = ..density..)) +
  geom_histogram(bins = 155, color = "green4", fill = "green2") + 
  geom_density(color = "darkgreen", lwd=1.3, bw = 70000)+
                 # geom_histogram(reads)
  labs(x = "Genome Position in Mb", y = "Density of reads mapped", caption = "Fig. 1: Density of the reads mapped to the E.coli genome.") +
  # scale_y_reverse(limits = c(0,12)) + 
  theme_classic() +
  theme(text = element_text(size = 20)) 
```
Description Fig x: Density of reads mapped to the E.coli genome. Each bin size represents a 30 kb long segment along the genome.

```{R}
ggplot(reads) +
  geom_histogram(data = reads[reads$strand == "s",],aes(x = pos/10^6, y = ..count..), bins = 155,color="lightblue",fill="blue") +
  # geom_density(data = reads[reads$strand == "s",], color="lightblue") +
  geom_histogram(data = reads[reads$strand == "a",],aes(x = pos/10^6, y = -..count..), bins = 155,color="orange",fill="red") +
  scale_y_continuous(labels = abs) + 
  labs(x = "Genome Position in Mb", y = "reads mapped per 30 Kb segment", caption = "Fig 2: Orientation of mapped Reads") +
  theme_classic() +
  theme(text = element_text(size = 20)) 
```
Description Fig x2: Sense vs. Antisense density of mapped reads. The y-axis shows a number of counts per 30 Kb segment. 

```{R}
ggplot(pie1, aes(x="", y=amount, fill=Group)) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0)+
  theme_void()+
  theme(legend.position="top") +
  theme(legend.title= element_blank()) +
  geom_text(aes(x = 1, y = ypos, label = amount), color = "white", size=5)+
  scale_fill_manual(values= c("blue", "red"))+
  labs(caption = "Fig 3: Number of successful mappings to sense and anti-sense strand.")
```
Description Fig x: Number of successful mappings to sense or anti-sense strand. 

```{R}
ggplot(pie2, aes(x="", y=value, fill=Group)) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0)+
  theme_void()+
  theme(legend.position="top") +
  theme(legend.title= element_blank()) +
  geom_text(aes(x = 1, y = ypos, label = value), color = "white", size=5)+
  scale_fill_manual(values= c("blue", "red"))+
labs(caption = "Fig 4: Number of reads that could be mapped to the gnome vs. number of reads that could not be mapped.")
```
Description Fig x: Number of reads that could be mapped to the gnome vs. number of reads that could not be mapped.
