#+title: Project Outline
#+description: Exam Project "Mapping the reads from NGS data to the genome"
#+authors: Abdullah Cetinkaya, Niklas Horner, Philip Wolper

Link to GitHub repository: https://github.com/pwolper/read-mapping

* Tasks to-do [3/4]
- [X] write/read results from respective folder (using os package?)
- [X] fix complementary strand search.
- [ ] Better visualizations. Read coverage with sliding window? How will this work efficiently?
- [X] Fix file naming and fix writing the data to a file. the csv file seems to be bugged a bit. Sorry if I shot it.


* Description
_“Mapping” the reads from NGS data to the genome_

- You have sequenced a genome with Illumina sequencing machine (single-end mode) and got a list of reads in FASTA file. For each read report the coordinate of the nucleotide where is has been mapped and to which strand it has been mapped.
- Please also keep in mind that DNA is double-stranded. In Fasta file of the genome you have only one strand reported, though the read can be mapped onto any of the strand. So please check both strands.
- Please provide the infographics showing how many reads have been mapped to the genome, how many have been mapped several times (how many times exactly?), how many haven’t been mapped.

* visualizations
Maybe add a plot showing coverage for regions on the genome. Sliding window approach?

* Resources

** NGS mapping from Tutorial for learning Genomics.
[[https://genomics.sschmeier.com/ngs-mapping/]]

** Groups
Book feedback session:
[[https://docs.google.com/document/d/1gqQZ-vmIGHv3nXg-H3GatdHKmJ-UM38COClTAJtsNO0/edit]]

Chat to find groups:
[[https://docs.google.com/document/d/1RknEWydbXxivWt1V83t9NufH-o0nXl4EqAnhhL4Eb0Y/edit]]

Project and student list:
[[https://docs.google.com/document/d/1nLNA5moYD3z_DvdHuh5t_v5nju3Gx5ferpy24_2UbcI/edit]]
