# Visualisation of mapped sequencing reads

```{R message=F}
library(ggplot2)
library(dplyr)
library(purrr)
library(here)

if (!require(ape)) {
  install.packages("ape")
  library(ape)
}
if (!require(ggrepel)) {
  install.packages("ggrepel")
  library(ggrepel)
}
if (!require(tidyverse)) {
  install.packages("tidyverse")
  library(tidyverse)
}

output <- here("output/img")
```

```{R}
genome <- read.dna(here("data/Ecoli_genome.fasta"),format="fasta")
sequence <- as.character(genome)
genome_length <- length(sequence)
```

## Importing data from /output/*.csv
```{R}
mapped <- read.csv(here("output/2023-01-27_NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome.csv"),sep = ";")

# mapped = mapped[50:60,]

mapped = mapped %>% mutate_at(vars(Sense, Antisense), ~map(strsplit(. ,split=","), ~map_int(.x, as.integer)))
# mapped %>% str()

sense = unlist(mapped$Sense)
anti = unlist(mapped$Antisense)

sense %>% length()
anti %>% length()

reads <- data.frame(pos = c(sense, anti), strand = as.factor(rep(c("s","a"),times = c(length(sense),length(anti)))))
```

```{R echo=FALSE}
pie1 <- data.frame(Group = c("Sense", "Anti-Sense"), amount = c(length(sense),length(anti)))
pie1 <- pie1 %>%
  arrange(desc(Group)) %>%
  mutate(ypos = c(pie1$amount[1]/2, pie1$amount[2]/2+pie1$amount[1])) #Calculate positions for labeling Pie chart. 
  
  #Pie chart: Hits vs No Hits
pie2 <- data.frame(Group = factor(c("Sense", "Anti-Sense", "Both", "No Hit")))
pie2$value = c(sum(mapped[,4] != 0 & mapped[,5] == 0),#count sense
               sum(mapped[,5] != 0 & mapped[,4] == 0),#count anti-sense
               sum(mapped[,4] != 0 & mapped[,5] != 0),#count both
               sum(mapped[,4] == 0 & mapped[,5] == 0))#count no hit
               

pie2$ypos <- c(pie2$value[1]/2+pie2$value[2]+pie2$value[3]+pie2$value[4],
               pie2$value[2]/2+pie2$value[3]+pie2$value[4],
               pie2$value[3]/2+pie2$value[4],
               pie2$value[4]/2) #Calculate positions for labeling Pie chart. 


  #Pie chart: Reads per Number of Hits

MapNum <- data.frame("Mappings" = 0, "Hits" = 0)


for(i in (1:length(mapped[,6]))){ #loop through "mapped" and check if number of total hits already in MapNum; if so add one to Hits; if not, add add number of toatal hits to new line.
    if (mapped[i,6] %in% MapNum[,1]){
      pos = which(MapNum[,1] == mapped[i,6])
      MapNum[pos,2] <- (MapNum[pos,2]+1)
      } else{
      newline <- c(mapped[i,6],1)
      MapNum <- rbind(MapNum, newline)
      }
}
MapNum$Fraction <- MapNum[,2]/sum(MapNum[,2])*100    #Calculate percentage
MapNum <- MapNum %>%                                 #Calculate positions for labeling Pie chart. 
  mutate(csum = rev(cumsum(rev(Hits))), 
         pos = Hits/2 + lead(csum, 1),
         pos = if_else(is.na(pos), Hits/2, pos))


```

## Plotting mapped reads
```{R}
p1 <- ggplot(reads, aes(x = pos, y = ..density..)) +
  geom_histogram(bins = 155, color = "green4", fill = "green2") + 
  geom_density(color = "darkgreen", lwd=1.3, bw = 70000)+
                 # geom_histogram(reads)
  labs(x = "Genome Position in Mb", y = "Density of reads mapped", caption = "Fig. 1: Density of the reads mapped to the E.coli genome.") +
  theme_classic()


p2 <- ggplot(reads) +
  geom_histogram(data = reads[reads$strand == "s",],aes(x = pos/10^6, y = ..count..), bins = 155,color="lightblue",fill="#0040ff") +
  # geom_density(data = reads[reads$strand == "s",], color="lightblue") +
  geom_histogram(data = reads[reads$strand == "a",],aes(x = pos/10^6, y = -..count..), bins = 155,color="orange",fill="#ff4000") +
  scale_y_continuous(labels = abs) + 
  labs(x = "Genome Position in Mb", y = "reads mapped per 30 Kb segment", caption = "Fig 2: Orientation of mapped Reads") +
  theme_classic() 

ggsave("F1_Genome_pos_total.tiff",p1, device = "tiff", dpi = 300, path = output)
ggsave("F2_Genome_pos_orient.tiff",p2, device = "tiff", dpi = 300, path = output) 
```
```{R fig_width = 100}
p1
p2
```


Description Fig 1: Density of reads mapped to the E.coli genome. Each bin size represents a 30 kb long segment along the genome.
Description Fig 2: Sense vs. Antisense density of mapped reads. The y-axis shows a number of counts per 30 Kb segment. 

```{R}
p3 <- ggplot(pie1, aes(x="", y=amount, fill=Group)) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)+
  theme_void()+
  theme(legend.position="top") +
  theme(legend.title= element_blank()) +
  geom_text(aes(x = 1, y = ypos, label = amount), color = "white", size=5)+
  scale_fill_manual(values= c("#ff4000","#0040ff"))+
  labs(caption = "Fig 3")

pie2$Group <- factor(pie2$Group, levels = c("Sense", "Anti-Sense", "Both", "No Hit"))

p4 <- ggplot(pie2, aes(x="", y=value, fill=Group)) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0)+
  theme_void()+
  theme(legend.position="top") +
  theme(legend.title= element_blank()) +
  geom_text(aes(x = 1.15, y = ypos, label = value), color = "white", size=5)+
  scale_fill_manual(values= c("#0040ff","#ff4000","green2","black"))+
  labs(caption = "Fig 4")

p5 <- ggplot(MapNum, aes(x="", y=Hits, fill=factor(Mappings))) +
  geom_bar(stat="identity", width=1, color = "black") +
  coord_polar("y", start=0)+
  theme_void()+
  theme(legend.position="top") +
  theme(legend.title= element_blank()) +
  scale_fill_brewer(name="Frequency of read",palette="Pastel1")+
  geom_label_repel(data = MapNum,
                   aes(y = pos, label = paste0(Hits, " (",Fraction, "%)")),
                   size = 5, nudge_x = 0.5, show.legend = FALSE) +
labs(caption = "Fig 5")

ggsave("F3_Hits_sense_vs_antisense.tiff",p3 ,device = "tiff", dpi = 300, path = output)
ggsave("F4_Read_distribution.tiff",p4, dpi = 300, path = output)
ggsave("F5_Reads_by_number_of_hits.png",p5 ,device = "png", dpi = 300, path = output)

```
```{R fig_width = 50}
p3
p4
p5
```

Description Fig 3: Number of successful hits to sense or anti-sense strand.
Description Fig 4: Number of reads mapped to sense strand, anti-sense strand, both strands or none in Genome
Description Fig 5: Frequency of hits per read. 
